src/server/pkg/storage/storage-format-spec.md

```diff
@@ -128,42 +132,135 @@ important differences between these two types of objects as it pertains to GC is
+
+* Always-running background process/service
+* No global locks or stoppages required during steady-state operations
+* Iterative and rate-limitable for any large operations
+* Batches updates from Pachyderm and batches deletes to object storage
+* Recovers smoothly from disorderly shutdown
+* Capable of becoming shardable with minimal changes
+
+### API
+
+The garbage collector will have two components, a client-side library and a
+service that runs in pachd.  The library will act as the single-point of contact
+for a client, while coordinating between the pachd service and the database as
+necessary.
+
+```go
+func (gcc *GcClient) ReserveChunks(chunks []Chunk)
+func (gcc *GcClient) UpdateReferences(add []Chunk, remove []Chunk)
+```
+
+The garbage collection service will provide a GRPC API for use by the client
+library (which should not be used directly).
+
+```
+rpc FlushDeletes(FlushDeletesRequest)
+```
+
+When deleting a chunk, the garbage collection service will need to evaluate the
+transitive references from that chunk to others.  Another service in the storage
+layer must provide an interface to fetch this information.
+
+```
+rpc GetReferences(GetReferencesRequest) stream Chunk
+
+message GetReferencesRequest {
+  Chunk chunk
+}
+```
+
+### Implementation
+
+**Client Library**
+
+`ReserveChunks` will store the set of chunks in new rows in the database with a
+preset TTL.  If the chunks are already present, the TTL will be updated.  If the
+chunks are currently being deleted, the client library will contact the pachd
+service and block until the deletion is confirmed and the chunk can be safely
+rewritten, using the `FlushDeletes` RPC.
+
+`UpdateReferences` will update the reference count of the given chunks in the
+database.  If the references drop to zero _and_ the chunk is outside of its
+TTL, the client library will call the `DeleteChunks` rpc on the pachd service.
+
+**Pachd Service**
+
+`FlushDeletes` is a synchronization RPC that will block until pending deletes
+for the specified chunks have completed.
+
+The service will periodically query the database for unreferenced chunks and
+proceed with removing their references and deleting them.  This is done by
+first marking the chunk as 'removing', then evaluating its references to other
+chunks through `GetReferences`.  Then, the service will atomically update these
+references in the database and mark the chunk as deleting.  Afterwards, it will
+send deletion requests to object storage, followed by deleting the row from the
+database, then responding to any `FlushDeletes` requests that were blocked.
+
+#### Persistence
+
+The garbage collector will save changes to a database which will act as the
+point of coordination between nodes.
+
+Database entry:
+* hash: string (chunk name, primary key)
+* refcount: number
+* updated: timestamp (used for TTL)
+* removing: timestamp (one-way)
+* deleting: timestamp (one-way)
+
+States:
+Nascent ↔ Live → Removing → Deleting → Deleted
```

Comment:
If two clients are trying to write the same chunk at the same time, what happens during the Nascent state in terms of reference counts and waiting/handling failures? I think `Removing` and `Deleting` can be combined if we decide not to go with resolving the transitive closure in the garbage collector. Could you elaborate on what the `Deleted` state could be used for?

Reply:
The Nascent state means that the chunk has no references (yet), but it is not to be removed until its TTL has passed.  Multiple clients upserting a Nascent chunk should only result in refreshing the TTL.

I don't think the garbage collector service needs to resolve the _transitive_ closure of referenced chunks, just the immediate references.  Any downstream chunks with zero references will be removed iteratively.  We could probably drop the `Removing` state if we are willing to put up with extra work that may be discarded in the case of a race condition.

The `Deleted` state is only here for explanation purposes, as it just means that the chunk is no longer present in object storage or the database.  There is no state associated with it aside from the lack of state.

Reply:
Drive-by comment: one case that has made garbage collection a difficult problem in the past is where a pipeline worker has finished processing a datum and has uploaded the output data to object store[1] but has not finished the output commit or uploaded the output commit's hashtree. It's hard for the garbage collector to know that it's safe to delete this chunk.

IIUC, in this proposal, this chunk should be in the `Nascent` state, but how should we handle refreshing the chunk's TTL while the pipeline is still running? Should workers continuously refresh the TTL of every chunk they've uploaded? What if the worker restarts?

[1] which I think we still do, even though the merger gets the filtered datum hashtree from the worker? In case of worker failure?

Reply:
> I don't think the garbage collector service needs to resolve the transitive closure of referenced chunks, just the immediate references. Any downstream chunks with zero references will be removed iteratively. We could probably drop the Removing state if we are willing to put up with extra work that may be discarded in the case of a race condition.

I think how this would actually play out is that a deletion of a FileSet, for example, would involve traversing to the base level, then making the appropriate `ReserveChunks` and `UpdateReferences` calls from the bottom up.

For the merge, it would be similar in that deletions would happen from the bottom up.

Reply:
@msteffen - I think that's a good reason to store semantic references inline in the database (one of the open questions).  If we can't accurately set a TTL for new chunks before they must be referenced, we could instead set a temporary semantic reference to that chunk, then remove it once the operation finishes.  If the temporary reference is in some way tied to the lifecycle of the client, we can periodically sweep invalid semantic references due to worker crashes.

I'll amend the proposal to include named (semantic) references in the database rows.

Reply:
@msteffen When a DatumSet (previously called chunks) is complete, it should be uploaded to object storage and I think the constituent chunks should not be in the `Nascent` state because they are live references (in that there is a semantic object that references them directly/indirectly). I think the core concern is the one I expressed earlier about a worker crashing while creating/modifying/deleting a FileSet. I think the solution @Tryneus is proposing is pretty similar to the one I proposed:

> At a high level, it seems like we would need to persist an identifier associated with a set of operations (maybe the semantic name of an object being created/modified/deleted) and the most recent chunk reference (the UpdateReferences operation could happen in a transaction and the last modified would be stored).

 but generalized to all of the chunks rather than the last one that was successfully modified. My thinking on this was that operations should in general be retried if they fail, but I think there are cases where we would not retry and we would not have enough information to know which chunks to cleanup without associating semantic information with each of the chunks. I think there is still an open question about how to know how far a worker got to prevent over/under counting references, and the solution I proposed does not work when the same chunk is directly/indirectly referenced twice by the semantic object being modified.